### PR TITLE
Use a single execution context per wasm interface (per thread) instead of per contract

### PR DESCRIPTION
**Pull request by linh2931**
## Change Description

A part of https://github.com/eosnetworkfoundation/product/issues/149.

Currently an execution context is created in a `backend` object for each contract, and cached (inside the backend) in `wasm_instantiation_cache` in the wasm interface object per thread. This wastes physical memory and prevents a single compiled WASM contract code shared across threads.

This PR provides an option to decouple execution context from backend, while keeps the existing behavior for backward compatibility. 

Resolves https://github.com/AntelopeIO/leap/issues/1456.

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: include/eosio/vm/backend.hpp [hunk 1]
```diff
@@ -74,49 +76,68 @@ namespace eosio { namespace vm {
...
       }
       // Leap:
       //  * Contract validation only needs single parsing as the instantiated module is not cached.
       //  * JIT execution needs single parsing only.
       //  * Interpreter execution requires two-passes parsing to prevent memory mappings exhaustion
-      backend(wasm_code_ptr& ptr, size_t sz, wasm_allocator* alloc, const Options& options = Options{}, bool single_parsing = true)
-         : memory_alloc(alloc), ctx(parse_module2(ptr, sz, options, single_parsing), detail::get_max_call_depth(options)) {
-         ctx.set_max_pages(detail::get_max_pages(options));
+      //  * Leap reuses execution context per thread; is_exec_ctx_created_by_backend is set
+      //  to false when a backend is constructued
```
Comment (heifner) [inline include/eosio/vm/backend.hpp:121]: As far as I know Leap is the only user of this submodule. Seems to me we should just change this to decouple exec ctx from backend. Supporting both modes, one of which is not used seems pointless.

**Review by heifner — CHANGES_REQUESTED** [review]

File: include/eosio/vm/backend.hpp [hunk 2]
```diff
@@ -74,49 +76,68 @@ namespace eosio { namespace vm {
...
+      backend(wasm_code_ptr& ptr, size_t sz, wasm_allocator* alloc, const Options& options = Options{}, bool single_parsing = true, bool is_exec_ctx_created_by_backend = true)
+         : memory_alloc(alloc), exec_ctx_created_by_backend(is_exec_ctx_created_by_backend), initial_max_call_depth(detail::get_max_call_depth(options)), initial_max_pages(detail::get_max_pages(options)) {
+         if (exec_ctx_created_by_backend) {
+            ctx = new context_t{parse_module2(ptr, sz, options, single_parsing), initial_max_call_depth};
```
[the remark this answers is not in the document]
Comment (linh2931) [reply include/eosio/vm/backend.hpp:125]: This same `ctx` pointer is assigned dynamically to the address of the exec ctx in the runtime object when `exec_ctx_created_by_backend` is false. That's why I did not use `std::unique_ptr`.

File: include/eosio/vm/backend.hpp (hunk 1 above)
Comment (linh2931) [reply include/eosio/vm/backend.hpp:121]: Thanks Kevin and Matt for this. I see `eos-vm` is submoduled in CDT `tools/external`. We can remove the dual modes (and make the code simpler) once decision on CDT is made.

Comment (heifner) [reply include/eosio/vm/backend.hpp:121]: Ok, sounds good.

**Review by heifner — APPROVED** [review]

**Pull request merged**